hazelcast/src/test/java/com/hazelcast/version/VersionTest.java

```diff
@@ -41,44 +42,49 @@
 public class VersionTest {
 
     private Version V3_0 = of(3, 0);
+    private Version NEGATIVE_VERSION_1 = of(-1, 0);
+    private Version NEGATIVE_VERSION_2 = of(-2, 0);
 
     @Test
-    public void getValue() throws Exception {
+    public void getValue() {
         assertEquals(3, V3_0.getMajor());
         assertEquals(0, V3_0.getMinor());
     }
 
     @Test
-    public void isEqualTo() throws Exception {
+    public void isEqualTo() {
         assertTrue(V3_0.isEqualTo(of(3, 0)));
         assertFalse(V3_0.isEqualTo(of(4, 0)));
+        assertTrue(NEGATIVE_VERSION_1.isEqualTo(of(-1, 0)));
     }
 
     @Test
-    public void isGreaterThan() throws Exception {
+    public void isGreaterThan() {
         assertTrue(V3_0.isGreaterThan(of(2, 0)));
         assertFalse(V3_0.isGreaterThan(of(3, 0)));
         assertFalse(V3_0.isGreaterThan(of(4, 0)));
+        assertTrue(V3_0.isGreaterThan(NEGATIVE_VERSION_1));
+        assertTrue(NEGATIVE_VERSION_1.isGreaterThan(NEGATIVE_VERSION_2));
```

Comment:
I'd expect this to be the other way around, when version is a WAN version.

I'm thinking about subclassing `Version`. That way we can create `ClusterVersion` and `WanVersion` that can
- compare/sort the versions properly,
- validate that once we set a cluster version, it can't be replaced by a WAN version (checking the sign bit) and vice-versa.

WDYT?

Reply:
Although we might argue that you need to know that negative values are also used as a special case for another kind of version to say that -2 should be greater than -1, I do agree this might be a bit confusing. Also, the fact that there is negative versions in the first place is confusing. I'll probably change this to only allow positive versions and restrict the negative version magic to only serialisation/deserialisation.